packages/medusa/src/services/payment-collection.ts

```diff
@@ -160,4 +189,493 @@ export default class PaymentCollectionService extends TransactionBaseService {
+        PaymentCollectionStatus.AUTHORIZED,
+        PaymentCollectionStatus.PARTIALLY_CAPTURED,
+        PaymentCollectionStatus.REQUIRES_ACTION,
+      ]
+
+      if (!allowedStatuses.includes(payCol.status)) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_ALLOWED,
+          `A Payment Collection with status ${payCol.status} cannot capture payment`
+        )
+      }
+
+      let captureError: Error | null = null
+      const capturedPayment = await this.paymentProviderService_
+        .withTransaction(manager)
+        .capturePayment(payment)
+        .catch((err) => {
+          captureError = err
+        })
+
+      payCol.captured_amount = payCol.captured_amount ?? 0
+      if (capturedPayment) {
+        payCol.captured_amount += payment.amount
+      }
+
+      if (payCol.captured_amount === 0) {
+        payCol.status = PaymentCollectionStatus.REQUIRES_ACTION
+      } else if (payCol.captured_amount === payCol.amount) {
+        payCol.status = PaymentCollectionStatus.CAPTURED
+      } else {
+        payCol.status = PaymentCollectionStatus.PARTIALLY_CAPTURED
+      }
+
+      const paymentCollectionRepository = manager.getCustomRepository(
+        this.paymentCollectionRepository_
+      )
+
+      await paymentCollectionRepository.save(payCol)
+
+      if (!capturedPayment) {
+        await this.eventBusService_
+          .withTransaction(manager)
+          .emit(PaymentCollectionService.Events.PAYMENT_CAPTURE_FAILED, {
+            ...payment,
+            error: captureError,
+          })
+
+        throw new MedusaError(
+          MedusaError.Types.UNEXPECTED_STATE,
+          `Failed to capture Payment ${payment.id}`
+        )
+      }
+
+      await this.eventBusService_
+        .withTransaction(manager)
+        .emit(PaymentCollectionService.Events.PAYMENT_CAPTURED, capturedPayment)
+
+      return capturedPayment
+    })
+  }
+
+  async capture(paymentId: string): Promise<Payment> {
+    const manager = this.transactionManager_ ?? this.manager_
+    const paymentCollectionRepository = manager.getCustomRepository(
+      this.paymentCollectionRepository_
+    )
+
+    const payCol =
+      await paymentCollectionRepository.getPaymentCollectionIdByPaymentId(
+        paymentId,
+        {
+          relations: ["payments"],
+        }
+      )
+
+    const payment = payCol.payments.find((payment) => paymentId === payment?.id)
+
+    return await this.capturePayment(payCol, payment!)
+  }
+
+  async captureAll(paymentCollectionId: string): Promise<Payment[]> {
+    const payCol = await this.retrieve(paymentCollectionId, {
+      relations: ["payments"],
+    })
+
+    const allPayments: Payment[] = []
+    for (const payment of payCol.payments) {
+      const captured = await this.capturePayment(payCol, payment).catch(
+        () => void 0
+      )
+
+      if (captured) {
+        allPayments.push(captured)
+      }
+    }
+
+    return allPayments
+  }
+
+  private async refundPayment(
```

Comment:
suggestion: Just spotted it, but it would be nice that private method are protected instead so that a user who would like to create a similar service extending this one have the flexibility to override those methods if needed. wdyt?